content/blog/2020-12-22-pipeline-II-analysis/index.md

```diff
@@ -0,0 +1,150 @@
+clock cycles required for memory read, the add operation, and the memory write.
+
+```c
+for (i = 1; i < N; i++)
+    mem[i] = mem[i-1] + i;
+```
+
+The overall minimum initiation interval, considering both resource constraints and data 
+dependences, is $MII = Max(ResMII, RecMII)$, which takes the maximum achievable II under both 
+circumstances. We can take a look at a warm-up example:
+
+```c
+int test()
+{
+  int A[20];
+  for (unsigned t = 2; t <= 20 - 1; t++)
+        A[t] = A[t - 2] + A[t - 1];
+  return 0;
+}
+```
+
+There are two load operations and one write operation in the loop body. If we assume the array 
+`A` is mapped to a RAM with one read port and one write port, then we can calculate ResMII to 
+be 2, as the number of memory read operations divided by the number of read ports. As for 
+loop-carried dependency, we notice that the current loop iteration needs to read array elements
+computed by the previous loop (`A[t - 1]`) and the one before the previous loop (`A[t - 2]`), 
+which means the operations in the current loop cannot get scheduled until `A[t - 2]` and `A[t - 1]`
+write back.  If we assume that the memory read and write operations both take 1 clock cycle, and 
+the add operation, since it is a combinational logic, can be merged with memory read into one clock 
+cycle, then the RecMII can be estimated as 2. The overall minimum initiation interval is $max(ResMII, 
+RecMII) = 2$.
+
+## Implementation
+I implemented an [LLVM pass](https://github.com/Sibylau/LLVM_10.0_walkthru/tree/master/llvm-pass/loop_II)
+to analyze the innermost loops. The problem is decomposed into two parts: estimate resource-constrained 
+ResMII and data-dependency bounded RecMII.
+
+For ResMII, it is easy to estimate using the formula (1). The pass traverses the loop, calculates 
+the number of independent loads and stores in the code, and divide it by the number of available 
+resources. I applied LLVM global value numbering pass on input programs before my handwritten pass, 
+in order to avoid counting recurring memory accesses. The number of resources, as the denominator of 
+the formula, is configured in advance as MACRO parameters. The pass I implemented only considered the
+memory resource constraints. There are other types of resources that can incur usage conflict, for 
+instance, the number of I/O ports, and certain compute resources, which I did not take into account 
+for the course of simplicity.
+
+The biggest challenge of this project is to estimate the RecMII, which basically requires precise 
+loop-carried dependency analysis. Ideally, I can use the maximum dependency distance as an approximate 
+of the RecMII. LLVM has a [loop dependency analysis pass](https://llvm.org/doxygen/DependenceAnalysis_8cpp_source.html). 
+It successfully extracted the dependency information in simple test programs, such as the warm-up 
+example above, but for some realistic benchmarks, it fails to figure out the dependence distance. 

```

Comment:
Makes sense—as with most nontrivial analyses, there's only so much complexity an analysis can handle and full precision will be unattainable. But it's nice when it works!